### DISCUSSION TITLE
How to mix serde `with =` and `serde_as` rules? / u128 as Hex

### DISCUSSION THREAD

**Discussion by steinerkelvin** [jonasbb/serde_with#507] [Q&A] [answered]
I would like to chain something like serde `with` field attribute and the rule `Hex` from `serde_with`:

```
#[serde_as]
#[derive(Debug, Serialize, Deserialize)]
struct Repr {
  #[serde_as(as = "serde_with::hex::Hex")]
  #[serde(with = "u128_as_buffer")]
  pub meta: u128,       // metadata
  
  ...
}

/// Serializes 
mod u128_as_buffer {
  type T = u128;
  pub fn serialize<S>(v: &T, s: S) -> Result<S::Ok, S::Error> where S: Serializer {
    todo!()
  }
  pub fn deserialize<'de, D>(d: D) -> Result<T, D::Error> where D: Deserializer<'de> {
    todo!()
  }
}
```

Can I achieve something like this?

Also, could there be a simple way to (de)serialize a number (in this case, an `u128`) as a hex string?

**Comment by jonasbb** [comment, marked as answer]
I am unsure what exactly you think the `u128_as_buffer` should be doing. But I am not aware of any way of chaining multiple functions. The `serde_as` macro will also not let you use `as` and `with` on the same field.

In your case, your module will need to do everything, i.e., turn the `u128` into a hex value and then instruct the `Serializer` how to serialize it. Of course it is also possible to use the `SerializeAs` trait from this crate, i.e., you can make a type similar to `Hex` which supports `u128`.

**Reply by steinerkelvin** [reply]
The macro [`serde_conv`](https://docs.rs/serde_with/2.0.0/serde_with/macro.serde_conv.html) seems pretty useful for this use case (`u128` <–> hex string).

**Reply by jonasbb** [reply]
I would not use `serde_conv` here, but really go for a full `DeserializeAs` implementation with `Visitor`. Of course, you can still macro it, if you need multiple implementation. Using a `Visitor` is not that complicated but you gain the benefits of using transient strings via `visit_str` (<https://serde.rs/lifetimes.html#transient-borrowed-and-owned-data>), saving some allocations, and using `expecting` for the error messages. You can see how [`DisplayFromStr`](https://docs.rs/serde_with/2.0.0/serde_with/struct.DisplayFromStr.html) is implemented.
